mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala

```diff
@@ -753,6 +782,39 @@ class SparseVector @Since("2.0.0") (
     }.unzip
     new SparseVector(selectedIndices.length, sliceInds.toArray, sliceVals.toArray)
   }
+
+  @Since("3.0.0")
+  override def iterator: Iterator[(Int, Double)] = {
```

Comment:
The complexity of traversal all elements is O(size), while the complexity of `var i = 0; while(i<vec.size) {val v = vec(i); i+=1}` is O(size * log(nnz))

Reply:
Can we just make this optimization rather than expose all these new API methods? I'm not clear we want to expose the 'iterator' vs 'foreachX' style. Why is this faster BTW?

Reply:
what about making them `private[spark]`?
I add the iterators, then I can make some impls (like `.ml.stat.Summarizer`) based on them, so that I can reuse it in both .ml and .mllib sides without vector conversion.

As to the performance gain of `SparseVector.iterator`/`SparseVector.foreach`, it is because that in this impl only numFeatures checks `if (i == k)` are needed to traserval all elements, while in the existing usages like in `RobustScaler`:
```scala
        var i = 0
        while (i < vec.size) {
          agg(i) = agg(i).insert(vec(i))
          i += 1
        }
```
numFeatures binarysearch are needed.

Reply:
I see, the idea is to reuse them within Spark? I'd say make them private and only the ones that are important for performance

Reply:
OK, I will make them private.
Other methods will make existing impl more concise in many places, for example:

in NaiveBayes:
```scala
 private[ml] def requireZeroOneBernoulliValues(v: Vector): Unit = {
    val values = v match {
      case sv: SparseVector => sv.values
      case dv: DenseVector => dv.values
    }

    require(values.forall(v => v == 0.0 || v == 1.0),
      s"Bernoulli naive Bayes requires 0 or 1 feature values but found $v.")
  }
```

will be changed to

```scala
private[ml] def requireZeroOneBernoulliValues(v: Vector): Unit = {
    require(v.nonZeroIterator.forall(v == 1.0),
      s"Bernoulli naive Bayes requires 0 or 1 feature values but found $v.")
}
```


in `VectorAssembler`:
```scala
vec.foreachActive { case (i, v) =>
          if (v != 0.0) {
            indices += featureIndex + i
            values += v
          }
        }
```
will be changed to
```scala
vec.foreachNonZero { case (i, v) =>
           indices += featureIndex + i
           values += v
       }
```

So I tend to also include other methods.